parser-typechecker/src/Unison/Util/Rope.hs

```diff
@@ -0,0 +1,170 @@
+{-# Language FunctionalDependencies #-}
+{-# Language DeriveFoldable #-}
+
+module Unison.Util.Rope where
+
+import Prelude hiding (drop,take,reverse,map,traverse)
+import Data.Foldable (toList)
+
+data Rope a
+  = Empty
+  | One {-# unpack #-} !Int {-# unpack #-} !a
```

Comment:
Probably not necessary to start with, but in the future may be nice to generalize over the measure to allow tracking things like newlines 😄 

Adds some annoying boilerplate in terms of a `Measured` typeclass and would likely hurt the ability to unpack; so what you have is probably best for now.